form/form.go

```diff
@@ -0,0 +1,191 @@
+package form
+
+import (
+	"bytes"
+	"net/url"
+	"reflect"
+	"strconv"
+)
+
+const tagName = "form"
+
+func AppendTo(values *RequestValues, i interface{}) {
+	//v := reflect.ValueOf(i).Elem()
+	reflectValue(values, reflect.ValueOf(i), nil)
+}
+
+func formatName(names []string) string {
+	if len(names) < 1 {
+		panic("Not allowed 0-length names slice")
+	}
+
+	fullName := names[0]
+	for i := 1; i < len(names); i++ {
+		fullName += "[" + names[i] + "]"
+	}
+	return fullName
+}
+
+func reflectValue(values *RequestValues, val reflect.Value, names []string) {
+	t := val.Type()
+
+	// Also do nothing if this is the type's zero value
+	if t.Comparable() && val.Interface() == reflect.Zero(t).Interface() {
+		return
+	}
+
+	switch val.Kind() {
+	case reflect.Array, reflect.Slice:
+		// formatName automatically adds square brackets, so just pass an empty
+		// string into the breadcrumb trail
+		arrNames := append(names, "")
+
+		for i := 0; i < val.Len(); i++ {
+			reflectValue(values, val.Index(i), arrNames)
+		}
+
+	case reflect.Bool:
+		values.Add(formatName(names), strconv.FormatBool(val.Bool()))
+
+	case reflect.Float32:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 32))
+
+	case reflect.Float64:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 64))
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		values.Add(formatName(names), strconv.FormatInt(val.Int(), 10))
+
+	case reflect.Interface:
+		if val.IsNil() {
+			return
+		}
+		reflectValue(values, val.Elem(), names)
+
+	case reflect.Map:
+		for _, keyVal := range val.MapKeys() {
+			if keyVal.Kind() != reflect.String {
+				panic("Don't support serializing maps with non-string keys")
+			}
+
+			reflectValue(values, val.MapIndex(keyVal), append(names, keyVal.String()))
+		}
+
+	case reflect.Ptr:
+		if val.IsNil() {
+			return
+		}
+		reflectValue(values, val.Elem(), names)
+
+	case reflect.String:
+		values.Add(formatName(names), val.String())
+
+	case reflect.Struct:
+		for i := 0; i < t.NumField(); i++ {
+			field := t.Field(i)
+			tag := field.Tag.Get(tagName)
+			if tag == "" {
+				continue
+			}
+			formName := tag
+			reflectValue(values, val.Field(i), append(names, formName))
+		}
+
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		values.Add(formatName(names), strconv.FormatUint(val.Uint(), 10))
+	}
+}
+
+// ---
+
+// RequestValues is a collection of values that can be submitted along with a
+// request that specifically allows for duplicate keys and encodes its entries
+// in the same order that they were added.
+type RequestValues struct {
```

Comment:
Unfortunately, I had to copy in the `RequestValues` struct from `params.go` for now, which makes this incompatible with our clients because these are of different values. Moving `RequestValues` will probably end up making this a breaking change, but hopefully not too many people will be depending on it directly.